common/src/main/java/com/microsoft/identity/common/internal/cache/WebAppsAccountIdRegistry.kt

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.internal.cache
+
+import androidx.annotation.VisibleForTesting
+import com.microsoft.identity.common.java.cache.IMultiTypeNameValueStorage
+import com.microsoft.identity.common.java.interfaces.IStorageSupplier
+import com.microsoft.identity.common.java.util.ObjectMapper
+import com.microsoft.identity.common.logging.Logger
+
+/**
+ * A registry that maps account IDs to sets of web app client IDs.
+ *
+ * This is used to track which web applications are associated with which accounts,
+ * enabling coordinated sign-out and management of web app sessions.
+ */
+class WebAppsAccountIdRegistry private constructor(
+    private val storage: IMultiTypeNameValueStorage
+){
+    private val lock = Any()
+
+    companion object {
+        private val TAG = WebAppsAccountIdRegistry::class.java.simpleName
+        private const val WEBAPPS_ACCOUNT_ID_REGISTRY_STORAGE_KEY = "WebAppsAccountIdRegistryStorageKey"
+        private const val MAP_JSON_KEY = "MapJsonKey"
+
+        /**
+         * Factory method to create an instance of [WebAppsAccountIdRegistry].
+         *
+         * @param supplier The storage supplier.
+         * @return A new instance of [WebAppsAccountIdRegistry].
+         */
+        fun create(supplier: IStorageSupplier): WebAppsAccountIdRegistry {
+            val store = supplier.getEncryptedFileStore(WEBAPPS_ACCOUNT_ID_REGISTRY_STORAGE_KEY)
+            return WebAppsAccountIdRegistry(store)
+        }
+    }
+
+    @Volatile
+    private var cache: MutableMap<String, MutableSet<String>>? = null
```

Comment:
if we only initialize one instance of this **_per request_** (not a static object), and only have 1 read/write at most **_per request_**, then having a cache wouldn't help much. (it'll just add complexity and overhead)

What's your use case?

Reply:
True, we are initializing this once per request, but I would say that the number of read/write will be more like 2 per request. For the sign out, we need to get all the client ids and then remove the account. For the getToken, I'm thinking it will be at least contains and potentially addClient.

Reply:
Lmk if a cache still seem like overkill for the usecases